### ISSUE TITLE
Recommend using `object` over `Any` for argument types?

### ISSUE THREAD

**Issue by srittau** [srittau/type-stub-pep#18]
This came up in python/typeshed#2539: Should we recommend using `object` instead of `Any` for argument types? From my understanding they are functionally equivalent, but `object` has no issues with any warning flags, does not require an import and is "more natural".

**Comment by Solumin** [comment]
I'm opposed to this. My main argument against it is just a strong feeling that `Any` exists for a reason, and why use `object` when `Any` exists? The warning flags for `Any` are Mypy's problem, other type checkers don't have those. The example given in the bug is the correct use case for `Any`. Not requiring an import doesn't seem like a very strong argument to me, since most type annotations are going to need to import some part of typing. And I definitely don't think `object` is "more natural" than `Any`, but that's just opinion anyway.

(Now, if you wanted to make `Any` an alias for `object`, I could be convinced of that.)

There's also the issue of stub files for Python 2 code. Should `object` match an instance of an old-style class? `Any` absolutely should, but `object` technically shouldn't. (On the other hand, how many libraries still using old-style classes would want to use type checking?)

There is a significant difference between `object` and `Any`: https://docs.python.org/3/library/typing.html#the-any-type
But I can see why this is a reason to use `object` over `Any`, since it stops users from annotating something as `Any` and then attempting to access attributes that may not exist.

**Comment by JelleZijlstra** [comment]
I mentioned something similar in the linked typeshed thread: `object` should be used only when literally any object is allowed. For example, the builtin `id()` takes an `object` argument.

**Comment by Solumin** [comment]
That makes sense to me: `object` is appropriate when the function takes explicitly an object qua object. That is, because it works directly with just an object and doesn't care about anything else.

**Comment by Michael0x2a** [comment]
FWIW, since I partially instigated this discussion...

In this particular case, I can concede I overshot. I suggested using `object` in the typeshed PR because it's the obvious alternative in most cases (for example, `EnumMeta.__contains__` actually does accept literally anything), and kept the suggested change to the contributions doc short to match the other bullet points. (I've amended my PR to be more detailed/nuanced a little while ago.)

Similarly, I guess there isn't necessarily a reason why `Enum.__new__` needed to be changed here vs adding some special-case in mypy -- it just turned out the typeshed change was the simplest solution in this case.

All that said, I do think it's worth writing docs in one form or another that nudge people away from using `Any` and point them towards alternatives they can try using instead. After all, each use of `Any` can potentially compromise the guarantee type checkers try and give you -- I feel pretty strongly that they're meant to be used exclusively as an escape hatch when it's just not possible to construct the correct type or when you must bridge typed and untyped code.

Of course, people can do whatever they want in their own codebases, but I do think it makes sense to steadily push shared resources like typeshed or third party stubs in a direction that minimizes the amount of dynamism they contain (and consequently reduces the amount of manual runtime checking end-users need to do).
